### ISSUE TITLE
duplicate key entry

### ISSUE THREAD

**Issue by joeck** [sapmentors/cds-pg#400]
I'm facing an issue where I have duplicate entries in my table with identical (composite) keys:

schema.cds
``` cds
entity CardReceiverAllocation {
    key receiver_ID         : UUID;
    key card_ID             : Integer;
    key timestamp           : Timestamp default $now; // this column was added later on
        receiver            : Association to one Receiver
                                  on receiver.ID = receiver_ID;
        card                : Association to one Card
                                  on card.cardID = card_ID;
        depositStatus       : Association to one CardDepositStatus;
        creatingUserEmail   : String(64);
        isCurrentAllocation : Boolean default false;
}

entity Card {
    key cardID      : Integer;
        isLost      : Boolean default false;
        pickupSlot  : Association to one PickupSlot;
        receiverAllocations : Association to many CardReceiverAllocation on
            cardID = receiverAllocations.card_ID;
        // [...]
}

entity Receiver : cuid { 
    firstName                : String(128);
    lastName                 : String(128);
    registrationDate         : Timestamp @cds.on.insert : $now;
    cardAllocations           : Association to many CardReceiverAllocation
                                   on  ID     = cardAllocations.receiver_ID;
    isBlocked                : Boolean default false;
    // [...]
}

aspect cuid {
    key ID : UUID;
}
```

The issue occurs in the CardReceiverAllocation table with the composite key of { receiver_ID, card_ID, timestamp } where the timestamp was added since multiple entries with same receiver and card are possible (a kind of history). When the timestamp column got added it was populated with default timestamp `now`. 

The problem is that we now have multiple rows with exact same key in the database (unique constraint violation):
<img width="1429" alt="image" src="https://github.com/sapmentors/cds-pg/assets/24994540/64dd1dbd-46b7-4d37-a6e9-a79093e39caf">
You can see that even the metadata uri is identical

Any idea on how to resolve this and why the unique constraint is not preventing this?

**Comment by ryegrosT8** [comment]
Hi there.

I guess that should be the columns names, the following associations at CardReceiverAllocation will automatically create two columns named “receiver_ID“ and “card_ID“ at database table CardReceiverAllocation : 

```
        receiver            : Association to one Receiver
                                  on receiver.ID = receiver_ID;
        card                : Association to one Card
                                  on card.cardID = card_ID;

```

But since you also create key columns with the same name, that surely is causing some problem. If the main purpose is to create some sort of log table, you can try to create the table with only one key... (GUID), like this:

``` 
entity CardReceiverAllocation {
    key ID              : UUID;
    timestamp           : Timestamp default $now; // this column was added later on
    receiver            : Association to one Receiver;
    card                : Association to one Card;
    depositStatus       : Association to one CardDepositStatus;
    creatingUserEmail   : String(64);
    isCurrentAllocation : Boolean default false;
}

entity Card {
    key cardID      : Integer;
        isLost      : Boolean default false;
        pickupSlot  : Association to one PickupSlot;
        receiverAllocations : Association to many CardReceiverAllocation on
            receiverAllocations.card = $self;
        // [...]
}

entity Receiver : cuid { 
    firstName                : String(128);
    lastName                 : String(128);
    registrationDate         : Timestamp @cds.on.insert : $now;
    cardAllocations           : Association to many CardReceiverAllocation
                                   on  cardAllocations.receiver = $self;
    isBlocked                : Boolean default false;
    // [...]
}
```

Is higly improbably that you get duplicate GUID’s at CardReceiverAllocation, unless you create millions of entries in small time interval, only in this case would be necessary to promote timestamp to key.

Cheers.
Rafael Yegros.

**Comment by joeck** [comment]
Hi Rafael

First of all thank you for your quick answer!

I thought specifiyng the association with `on receiver.ID = receiver_ID` counts as a "unmanaged association" which will not automatically create the `_ID` fields as we specify the foreign key ourselves. Or am I misunderstanding something? See explanation here: https://cap.cloud.sap/docs/guides/domain-modeling#managed-1-associations
![image](https://github.com/sapmentors/cds-pg/assets/24994540/c0b7b567-6979-4be2-99f7-5882187b8365)

Since we added the `timestamp` column later on when the `CardReceiverAllocation` table already had data and we wanted the `timestamp` field as part of the key all the existing rows got the timestamp (`default $now`) in a short timeframe. In the database we actually have over 1300 entries with the exact same timestamp from when we populated the colum and it got filled with the `default $now`. Maybe because the `$now` gets translated to an exact timestamp before inserting? Or the Database just being so quick? But wouldn't this throw a unique constraint violation if that resulted in rows with duplicate GUID's?
